File: rust/day9_part2_2019.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::collections::VecDeque;

struct IntcodeComputer {
    memory: Vec<i64>,
    ip: usize,
    relative_base: i64,
    input: VecDeque<i64>,
    output: Vec<i64>,
}

impl IntcodeComputer {
    fn new(memory: Vec<i64>) -> IntcodeComputer {
        IntcodeComputer {
            memory,
            ip: 0,
            relative_base: 0,
            input: VecDeque::new(),
            output: Vec::new(),
        }
    }

    fn run(&mut self) {
        loop {
            let opcode = self.memory[self.ip] % 100;
            let mode1 = (self.memory[self.ip] / 100) % 10;
            let mode2 = (self.memory[self.ip] / 1000) % 10;
            let mode3 = (self.memory[self.ip] / 10000) % 10;

            match opcode {
                1 => {
                    let a = self.get_param(mode1, self.ip + 1);
                    let b = self.get_param(mode2, self.ip + 2);
                    let c = self.get_addr(mode3, self.ip + 3);
                    self.memory[c as usize] = a + b;
                    self.ip += 4;
                }
                2 => {
                    let a = self.get_param(mode1, self.ip + 1);
                    let b = self.get_param(mode2, self.ip + 2);
                    let c = self.get_addr(mode3, self.ip + 3);
                    self.memory[c as usize] = a * b;
                    self.ip += 4;
                }
                3 => {
                    let a = self.get_addr(mode1, self.ip + 1);
                    self.memory[a as usize] = self.input.pop_front().unwrap_or(0);
                    self.ip += 2;
                }
                4 => {
                    let a = self.get_param(mode1, self.ip + 1);
                    self.output.push(a);
                    self.ip += 2;
                }
                5 => {
                    let a = self.get_param(mode1, self.ip + 1);
                    let b = self.get_param(mode2, self.ip + 2);
                    if a != 0 {
                        self.ip = b as usize;
                    } else {
                        self.ip += 3;
                    }
                }
                6 => {
                    let a = self.get_param(mode1, self.ip + 1);
                    let b = self.get_param(mode2, self.ip + 2);
                    if a == 0 {
                        self.ip = b as usize;
                    } else {
                        self.ip += 3;
                    }
                }
                7 => {
                    let a = self.get_param(mode1, self.ip + 1);
                    let b = self.get_param(mode2, self.ip + 2);
                    let c = self.get_addr(mode3, self.ip + 3);
                    if a < b {
                        self.memory[c as usize] = 1;
                    } else {
                        self.memory[c as usize] = 0;
                    }
                    self.ip += 4;
                }
                8 => {
                    let a = self.get_param(mode1, self.ip + 1);
                    let b = self.get_param(mode2, self.ip + 2);
                    let c = self.get_addr(mode3, self.ip + 3);
                    if a == b {
                        self.memory[c as usize] = 1;
                    } else {
                        self.memory[c as usize] = 0;
                    }
                    self.ip += 4;
                }
                9 => {
                    let a = self.get_param(mode1, self.ip + 1);
                    self.relative_base += a;
                    self.ip += 2;
                }
                99 => {
                    break;
                }
                _ => {
                    panic!("Unknown opcode");
                }
            }
        }
    }

    fn get_param(&self, mode: i64, index: usize) -> i64 {
        match mode {
            0 => self.memory[self.memory[index] as usize],
            1 => self.memory[index],
            2 => self.memory[(self.relative_base + self.memory[index]) as usize],
            _ => panic!("Unknown mode"),
        }
    }

    fn get_addr(&self, mode: i64, index: usize) -> i64 {
        match mode {
            0 => self.memory[index] as i64,
            2 => self.relative_base + self.memory[index],
            _ => panic!("Unknown mode"),
        }
    }
}

fn main() -> io::Result<()> {
    let file = File::open("input.txt")?;
    let reader = io::BufReader::new(file);
    let mut memory: Vec<i64> = reader
        .lines()
        .next()
        .unwrap()?
        .split(',')
        .map(|x| x.parse().unwrap())
        .collect();

    memory.resize(10000, 0);

    let mut computer = IntcodeComputer::new(memory);
    computer.input.push_back(2); // Run in sensor boost mode
    computer.run();

    println!("Distress signal coordinates: {}", computer.output[0]);

    Ok(())
}

File: elixir/day9_part2_2019.exs

defmodule IntcodeComputer do
  def run(program, input) do
    do_run(program, 0, 0, input, [])
  end

  defp do_run(program, pc, relative_base, input, output) do
    opcode = program[pc] |> rem(100)
    mode1 = div(program[pc], 100) |> rem(10)
    mode2 = div(program[pc], 1000) |> rem(10)
    mode3 = div(program[pc], 10000) |> rem(10)

    case opcode do
      1 ->
        p1 = get_value(program, pc + 1, mode1, relative_base)
        p2 = get_value(program, pc + 2, mode2, relative_base)
        p3 = get_address(program, pc + 3, mode3, relative_base)
        do_run(Map.put(program, p3, p1 + p2), pc + 4, relative_base, input, output)

      2 ->
        p1 = get_value(program, pc + 1, mode1, relative_base)
        p2 = get_value(program, pc + 2, mode2, relative_base)
        p3 = get_address(program, pc + 3, mode3, relative_base)
        do_run(Map.put(program, p3, p1 * p2), pc + 4, relative_base, input, output)

      3 ->
        [h | t] = input
        p1 = get_address(program, pc + 1, mode1, relative_base)
        do_run(Map.put(program, p1, h), pc + 2, relative_base, t, output)

      4 ->
        p1 = get_value(program, pc + 1, mode1, relative_base)
        do_run(program, pc + 2, relative_base, input, [p1 | output])

      5 ->
        p1 = get_value(program, pc + 1, mode1, relative_base)
        p2 = get_value(program, pc + 2, mode2, relative_base)

        if p1 != 0 do
          do_run(program, p2, relative_base, input, output)
        else
          do_run(program, pc + 3, relative_base, input, output)
        end

      6 ->
        p1 = get_value(program, pc + 1, mode1, relative_base)
        p2 = get_value(program, pc + 2, mode2, relative_base)

        if p1 == 0 do
          do_run(program, p2, relative_base, input, output)
        else
          do_run(program, pc + 3, relative_base, input, output)
        end

      7 ->
        p1 = get_value(program, pc + 1, mode1, relative_base)
        p2 = get_value(program, pc + 2, mode2, relative_base)
        p3 = get_address(program, pc + 3, mode3, relative_base)

        if p1 < p2 do
          do_run(Map.put(program, p3, 1), pc + 4, relative_base, input, output)
        else
          do_run(Map.put(program, p3, 0), pc + 4, relative_base, input, output)
        end

      8 ->
        p1 = get_value(program, pc + 1, mode1, relative_base)
        p2 = get_value(program, pc + 2, mode2, relative_base)
        p3 = get_address(program, pc + 3, mode3, relative_base)

        if p1 == p2 do
          do_run(Map.put(program, p3, 1), pc + 4, relative_base, input, output)
        else
          do_run(Map.put(program, p3, 0), pc + 4, relative_base, input, output)
        end

      9 ->
        p1 = get_value(program, pc + 1, mode1, relative_base)
        do_run(program, pc + 2, relative_base + p1, input, output)

      99 ->
        Enum.reverse(output)

      _ ->
        raise "Unknown opcode: #{opcode}"
    end
  end

  defp get_value(program, address, mode, relative_base) do
    case mode do
      0 -> program[program[address] || 0] || 0
      1 -> program[address] || 0
      2 -> program[(program[address] || 0) + relative_base] || 0
    end
  end

  defp get_address(program, address, mode, relative_base) do
    case mode do
      0 -> program[address] || 0
      2 -> (program[address] || 0) + relative_base
      _ -> raise "Invalid address mode for write operation: #{mode}"
    end
  end
end

input =
  File.read!("input.txt")
  |> String.trim()
  |> String.split(",")
  |> Enum.map(&String.to_integer/1)
  |> Enum.with_index()
  |> Map.new(fn {val, idx} -> {idx, val} end)

# Part 1
output1 = IntcodeComputer.run(input, [1])
IO.puts("BOOST keycode (Part 1): #{Enum.join(output1, ",")}")

# Part 2
output2 = IntcodeComputer.run(input, [2])
IO.puts("Distress signal coordinates (Part 2): #{Enum.join(output2, ",")}")
